### PR TITLE
Upgrade dep moduels to 0.24.0

### PR DESCRIPTION
**Pull request by zmvictor**
### Changes
- upgrade core modules to 0.24.0
- fix bug in codegen script: run codegen right after deleting the stale generated files, otherwise dep update will fail due to required pkg is deleted
- update clientset in frameworkcontroller repo to use new interfaces
- update docker file 

### Tests
```
$ ./hack/update-all.sh
$ ./build/frameworkcontroller/go-build.sh
$ ./build/frameworkcontroller/docker-build.sh
$ ./build/frameworkbarrier/go-build.sh
$ ./build/frameworkbarrier/docker-build.sh
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,20 +1,60 @@
 module github.com/microsoft/frameworkcontroller
 
-go 1.12
+go 1.18
 
 require (
-	github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da // indirect
+	github.com/pkg/errors v0.9.1
```
Comment (yqwang-ms) [inline go.mod:6]: Why much more dependencies now? Expected? Could we remove some?

File: go.mod [hunk 2]
```diff
@@ -1,20 +1,60 @@
 module github.com/microsoft/frameworkcontroller
 
-go 1.12
+go 1.18
```
Comment (yqwang-ms) [inline go.mod:4]: What if still use 1.12? Change to 1.18 may break 1.12 builders?

File: build/frameworkbarrier/docker-build.sh [hunk 3]
```diff
@@ -32,6 +32,6 @@ IMAGE_NAME=frameworkbarrier
 
 cd ${PROJECT_DIR}
 
-docker build -t ${IMAGE_NAME} -f ${BASH_DIR}/Dockerfile .
+docker build --network host -t ${IMAGE_NAME} -f ${BASH_DIR}/Dockerfile .
```
Comment (yqwang-ms) [inline build/frameworkbarrier/docker-build.sh:35]: >--network host [](http://example.com/codeflow?start=13&length=14)

Why add "--network host"? Is it a must?

File: go.mod (hunk 2 above)
Comment (yqwang-ms) [reply go.mod:4]: Can we keep 1.12 in this PR, and this PR only focus on upgrade k8s packages?

File: hack/update-all.sh [hunk 4]
```diff
@@ -34,12 +34,12 @@ GOPKG_LOCK_FILE=${PROJECT_DIR}/Gopkg.lock
 
 cd ${PROJECT_DIR}
 
-# Remove possible stale import in generated code
-rm -rf ${CODEGEN_DIR} ${CODEGEN_FILE} ${GOPKG_LOCK_FILE}
-
 # Update dependent package for non-generated code, such as the code-generator itself
 ${BASH_DIR}/update-dep.sh
```
Comment (yqwang-ms) [inline hack/update-all.sh:39]: Why move this?  We should first clean up generated files, then update dependent package.

File: pkg/apis/frameworkcontroller/v1/zz_generated.deepcopy.go [hunk 5]
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated
```
Comment (yqwang-ms) [inline pkg/apis/frameworkcontroller/v1/zz_generated.deepcopy.go:2]: Duplicated?

File: pkg/barrier/barrier.go [hunk 6]
```diff
@@ -225,6 +227,7 @@ func NewFrameworkBarrier() *FrameworkBarrier {
 
 func (b *FrameworkBarrier) Run() {
 	klog.Infof("Running %v", ComponentName)
+	ctx := context.TODO()
```
Comment (yqwang-ms) [inline pkg/barrier/barrier.go:230]: >context.TODO() [](http://example.com/codeflow?start=8&length=14)

Why previously use the context.Background, and now changed to context.TODO? Are they both work? What is the difference?

File: pkg/barrier/barrier.go [hunk 7]
```diff
@@ -236,7 +239,7 @@ func (b *FrameworkBarrier) Run() {
 		func() (bool, error) {
 			f, err = b.fClient.FrameworkcontrollerV1().
 				Frameworks(b.bConfig.FrameworkNamespace).
-				Get(b.bConfig.FrameworkName, meta.GetOptions{})
+				Get(ctx, b.bConfig.FrameworkName, meta.GetOptions{})
```
Comment (yqwang-ms) [inline pkg/barrier/barrier.go:242]: >ctx [](http://example.com/codeflow?start=8&length=3)

Just "context.TODO()"?

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -1288,8 +1290,11 @@ func (c *FrameworkController) deleteFramework(
 		"[%v]: Failed to delete Framework %v: confirm: %v: ",
 		f.Key(), f.UID, confirm)
 
+	ctx := context.TODO()
 	deleteErr := c.fClient.FrameworkcontrollerV1().Frameworks(f.Namespace).Delete(
-		f.Name, &meta.DeleteOptions{
+		ctx,
```
Comment (yqwang-ms) [inline pkg/controller/controller.go:1295]: >ctx [](http://example.com/codeflow?start=2&length=3)

I checked K8s offical controllers, seems ctx no need to reuse, so here just "context.TODO"?

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -1288,8 +1290,11 @@ func (c *FrameworkController) deleteFramework(
...
+		f.Name,
+		meta.DeleteOptions{
```
Comment (yqwang-ms) [inline pkg/controller/controller.go:1297]: merge to same line?

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -2354,7 +2366,7 @@ func (c *FrameworkController) updateRemoteFrameworkStatus(f *ci.Framework) error
 			}
 		}
 
-		_, updateErr := c.fClient.FrameworkcontrollerV1().Frameworks(updateF.Namespace).Update(updateF)
+		_, updateErr := c.fClient.FrameworkcontrollerV1().Frameworks(updateF.Namespace).Update(ctx, updateF, meta.UpdateOptions{})
 		return updateErr
```
Comment (yqwang-ms) [inline pkg/controller/controller.go:2370]: break to 2 lines? like
```
  _, updateErr := c.fClient.FrameworkcontrollerV1().Frameworks(updateF.Namespace).Update(
ctx, updateF, meta.UpdateOptions{})

````

File: pkg/internal/utils.go [hunk 11]
```diff
@@ -37,20 +41,18 @@ import (
...
-	"reflect"
-	"time"
 )
 
 func CreateClients(kConfig *rest.Config) (
 	kubeClient.Interface, frameworkClient.Interface) {
 	kClient, err := kubeClient.NewForConfig(kConfig)
 	if err != nil {
-		panic(fmt.Errorf("Failed to create KubeClient: %v", err))
+		panic(fmt.Errorf("failed to create KubeClient: %v", err))
```
Comment (yqwang-ms) [inline pkg/internal/utils.go:50]: >failed [](http://example.com/codeflow?start=20&length=6)

Keep orignal logging style?

File: go.mod (hunk 1 above)
Comment (zmvictor) [reply go.mod:6]: This is code-gen from `go mod tidy`. Let me see if there is less

File: pkg/controller/controller.go (hunk 8 above)
Comment (zmvictor) [reply pkg/controller/controller.go:1295]: The follow up change may either pass in a ctx or create one as above lines. Both of them will keep ctx in following usage so I will leave as current version.

File: build/frameworkbarrier/docker-build.sh (hunk 3 above)
Comment (zmvictor) [reply build/frameworkbarrier/docker-build.sh:35]: w/o this change, I encountered:
```
ERROR: https://dl-cdn.alpinelinux.org/alpine/v3.15/main: temporary error (try again later)
WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.15/main: No such file or directory
ERROR: https://dl-cdn.alpinelinux.org/alpine/v3.15/community: temporary error (try again later)
WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.15/community: No such file or directory
```

File: hack/update-all.sh (hunk 4 above)
Comment (zmvictor) [reply hack/update-all.sh:39]:
```
github.com/microsoft/frameworkcontroller/pkg/barrier imports
        github.com/microsoft/frameworkcontroller/pkg/client/clientset/versioned: no required module provides package github.com/microsoft/frameworkcontroller/pkg/client/clientset/versioned; to add it:
        go get github.com/microsoft/frameworkcontroller/pkg/client/clientset/versioned
github.com/microsoft/frameworkcontroller/pkg/controller imports
        github.com/microsoft/frameworkcontroller/pkg/client/informers/externalversions: no required module provides package github.com/microsoft/frameworkcontroller/pkg/client/informers/externalversions; to add it:
        go get github.com/microsoft/frameworkcontroller/pkg/client/informers/externalversions
github.com/microsoft/frameworkcontroller/pkg/controller imports
        github.com/microsoft/frameworkcontroller/pkg/client/listers/frameworkcontroller/v1: no required module provides package github.com/microsoft/frameworkcontroller/pkg/client/listers/frameworkcontroller/v1; to add it:
        go get github.com/microsoft/frameworkcontroller/pkg/client/listers/frameworkcontroller/v1
```

**Comment by yqwang-ms** [comment]
Close the stale PR

**Pull request closed without merging**